sdks/python/apache_beam/yaml/yaml_transform.py

```diff
@@ -208,6 +220,56 @@ def get_outputs(self, transform_name):
   def compute_outputs(self, transform_id):
     return expand_transform(self._transforms_by_uuid[transform_id], self)
 
+  def best_provider(self, t, input_providers):
```

Comment:
This function is pretty complicated and uses a lot of nested lists/dicts. It's hard to keep track of what is happening without types and defined structures. Would you agree with this? Can we perhaps plan some work to refactor it a little bit?

Reply:
It helps, thanks